index.js

```diff
@@ -142,29 +144,45 @@ GeneratePackageJsonPlugin.prototype.apply = function(compiler) {
       }
     });
 
-    // Overwrite modules or set new module dependencies for those that have been
-    // deliberately set in " otherPackageValues.dependencies "
-    if (this.otherPackageValues && this.otherPackageValues.dependencies) {
-      const nonWebpackModuleNames = Object.keys(this.otherPackageValues.dependencies);
-
-      for (let k = 0; k < nonWebpackModuleNames.length; k += 1) {
-        const moduleName = nonWebpackModuleNames[k];
-
-        if (this.otherPackageValues.dependencies[moduleName] && this.otherPackageValues.dependencies[moduleName].length > 0) {
-          logIfDebug(`GPJWP: Adding deliberate module with version set deliberately: ${moduleName} -> ${this.otherPackageValues.dependencies[moduleName]}`);
-          modules[moduleName] = this.otherPackageValues.dependencies[moduleName];
-        } else if (this.dependencyVersionMap[moduleName] != null) {
-          logIfDebug(`GPJWP: Adding deliberate module with version found in sources: ${moduleName} -> ${this.dependencyVersionMap[moduleName]}`);
-          modules[moduleName] = this.dependencyVersionMap[moduleName];
-        } else {
-          console.warn(`GeneratePackageJsonPlugin: You have set a module to be included deliberately with name: "${moduleName}" - but there is no version specified in any source files!`);
+    const basePackageValues = Object.assign({}, this.otherPackageValues);
+
+    for (let i = 0; i < dependencyTypes.length; i += 1) {
+      const dependencyType = dependencyTypes[i];
+
+      // Overwrite modules or set new module dependencies for those that have been
+      // deliberately set in " basePackageValues.[dependencyType] "
+      // This mechanism ensures that dependencies declared in the basePackageValues
+      // take precedence over the found dependencies
+      if (basePackageValues && basePackageValues[dependencyType]) {
+        const nonWebpackModuleNames = Object.keys(basePackageValues[dependencyType]);
+
+        for (let k = 0; k < nonWebpackModuleNames.length; k += 1) {
+          const moduleName = nonWebpackModuleNames[k];
+
+          if (basePackageValues[dependencyType][moduleName] && basePackageValues[dependencyType][moduleName].length > 0) {
```

Comment:
This might throw an undefined error because the `basePackageValues` might not always include all the dependency types. I think we could use the newer optional chaining syntax to get around that, potentially:

```js
if ((basePackageValues[dependencyType]?.[moduleName]?.length ?? 0)) > 0)
```

Reply:
Actually rather not the optional chaining, if people are using older Node.js versions (13 and below) it will throw.

Will have to do the traditional:

```js
if (basePackageValues[dependencyType] && basePackageValues[dependencyType][moduleName] && basePackageValues[dependencyType][moduleName].length > 0)
```